onnxruntime/core/providers/cpu/tensor/concat.cc

```diff
@@ -17,59 +17,82 @@ Status ConcatBase::PrepareForCompute(OpKernelContext* ctx, int input_count, Prep
   const Tensor* tensor_pointer = ctx->Input<Tensor>(0);
   if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
   const Tensor& inputs_0 = *tensor_pointer;
+  const auto& inputs_0_dims = inputs_0.Shape().GetDims();
+  const size_t inputs_0_rank = inputs_0_dims.size();
+  ORT_RETURN_IF_NOT(inputs_0_rank > 0, "Cannot concatenate scalars");
 
   auto axis = HandleNegativeAxis(axis_, inputs_0.Shape().NumDimensions());
 
+  // cache num of elements in tensor for later use
+  // as it's expensive to call Size() on TensorShape over and over
+  std::vector<size_t> tensor_num_elements(input_count);
   // Ensure all of the non concatenated axes match each other
   for (int index = 1; index < input_count; index++) {
+    size_t num_elements = 1;
     tensor_pointer = ctx->Input<Tensor>(index);
     if (tensor_pointer == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
-    auto& data_n = *tensor_pointer;
-    // Ensure all the other axes match
-    auto dimension_count = inputs_0.Shape().NumDimensions();
-    for (int axis_index = 0; axis_index < dimension_count; axis_index++) {
+    auto& inputs_n = *tensor_pointer;
+    const auto& inputs_n_dims = inputs_n.Shape().GetDims();
+    const size_t inputs_n_rank = inputs_n_dims.size();
+    ORT_ENFORCE(inputs_n_rank == inputs_0_rank, "Ranks of input data are different, cannot concatenate them");
+    // Ensure all the other (non-concat) axes match
+    for (int axis_index = 0; axis_index < inputs_0_rank; axis_index++) {
```

Comment:
prefer using ++axis_index as a habit, although it doesn't matter here so much

Reply:
This is from the earlier implementation as is - but will fix